### PR TITLE
Fix reapTime logic in NetworkDB

### PR DESCRIPTION
**Pull request by fcrisciani**
- Added remainingReapTime field in the table event.
  Wihtout it a node that did not have a state for the element
  was marking the element for deletion setting the max reapTime.
  This was creating the possibility to keep the entry being resync
  between nodes forever avoding the purpose of the reap time
  itself.

- On broadcast of the table event the node owner was rewritten
  with the local node name, this was not correct because the owner
  should continue to remain the original one of the message

Signed-off-by: Flavio Crisciani <flavio.crisciani@docker.com>

### CODE REVIEW

File: networkdb/cluster.go [hunk 1]
```diff
@@ -321,43 +321,49 @@ func (nDB *NetworkDB) reapNetworks() {
...
 
-		if _, ok := nDB.indexes[byNetwork].Delete(fmt.Sprintf("/%s/%s/%s", nid, tname, key)); !ok {
-			logrus.Errorf("Could not delete entry in network %s with table name %s and key %s as it does not exist", nid, tname, key)
-		}
+			if entry.reapTime > reapPeriod+timeCompensation {
+				entry.reapTime -= reapPeriod + timeCompensation
+				return false
+			}
+
+			params := strings.Split(path[1:], "/")
```
Comment (vieux) [inline networkdb/cluster.go]: are we 100% sure the path will have 4 `/` in it ?

File: networkdb/cluster.go [hunk 2]
```diff
@@ -321,43 +321,49 @@ func (nDB *NetworkDB) reapNetworks() {
...
+			nid := params[0]
+			tname := params[1]
+			key := params[2]
+
+			okTable, okNetwork := nDB.deleteEntry(nid, tname, key)
+			if !okTable {
+				logrus.Errorf("Could not delete entry in table %s with network id %s and key %s as it does not exist", tname, nid, key)
```
Comment (vieux) [inline networkdb/cluster.go]: I'm having some trouble reading the message, could we replace `it` by `the table` here and `the network` a couple of lines down ?

File: networkdb/cluster.go (hunk 1 above)
Comment (fcrisciani) [reply networkdb/cluster.go]: why 4? it is actually 3 like: `fmt.Sprintf("/%s/%s/%s", nid, tname, key)`
Also the insertion and deletion is now contained in single functions: createOrUpdateEntry and deleteEntry

Comment (vieux) [reply networkdb/cluster.go]: sorry I meant 3

File: networkdb/cluster.go (hunk 2 above)
Comment (fcrisciani) [reply networkdb/cluster.go]: I actually left the original messages, I just had to change the bool on top. 
Anyway here `it` refers to the entry itself not the table.
Maybe a better rephrase can be: Table delete failed, entry with key:%s does not exists in the table:%s network:%s
WDYT?

File: networkdb/delegate.go [hunk 3]
```diff
@@ -235,20 +234,26 @@ func (nDB *NetworkDB) handleTableEvent(tEvent *TableEvent) bool {
...
+		reapTime: time.Duration(tEvent.ResidualReapTime) * time.Second,
 	}
 
-	if e.deleting {
+	// All the entries marked for deletion should have a reapTime set greater than 0
+	// This case can happen if the cluster is running different versions of the engine where the old version does not have the
+	// field. If that is not the case, this can be a BUG
+	if e.deleting && e.reapTime == 0 {
+		logrus.Warnf("handleTableEvent object %+v has a 0 reapTime, is the cluster running the same docker engine version?", tEvent)
 		e.reapTime = reapInterval
```
Comment (vieux) [inline networkdb/delegate.go]: is this change of behavior expected here ? 
shouldn't we log when `e.reapTime == 0` but still do `e.reapTime = reapInterval` if it is not ?

File: networkdb/cluster.go (hunk 2 above)
Comment (vieux) [reply networkdb/cluster.go]: @fcrisciani yeah, sound good

File: networkdb/delegate.go (hunk 3 above)
Comment (fcrisciani) [reply networkdb/delegate.go]: nope, this is actually the point of this patch, the reapTime should be the one passed in the proto (line 237).
If that is 0 and the entry is marked for deletion we are in front of a bug (nodes does not set the value properly) or we are dealing with an old client that does not set it because does not have the new proto.

The old behavior was such that no matter when you get the notification the clock was being set to the max value of 30 min

File: networkdb/cluster.go (hunk 2 above)
Comment (fcrisciani) [reply networkdb/cluster.go]: k changing it

**Comment by vieux** [comment]
LGTM (with the log format update)

File: networkdb/delegate.go (hunk 3 above)
Comment (mavenugo) [reply networkdb/delegate.go]: Except in that very unfortunate circumstance that the ResidualReapTime populated by the peer has infact gone down to exactly 0 seconds at the time of populating the message (https://github.com/docker/libnetwork/pull/1944/files#diff-ced1ed36c3141cb15bdf5acafed78d67R583). If that happens in such a rare situation, we will reset this back to 30 seconds and the event will get cleaned up in 30 seconds (or in the more rarest of occasions we get another genuine 0 seconds ResidualReapTime). 

@fcrisciani is that correct ?

Comment (mavenugo) [reply networkdb/delegate.go]: BTW. yes. This is the actual point of this patch and I think it is a good idea to get this done.

File: networkdb/delegate.go [hunk 4]
```diff
@@ -303,22 +308,17 @@ func (nDB *NetworkDB) handleTableMessage(buf []byte, isBulkSync bool) {
...
 		}
 
-		broadcastQ.QueueBroadcast(&tableEventMessage{
+		n.tableBroadcasts.QueueBroadcast(&tableEventMessage{
 			msg:   buf,
 			id:    tEvent.NetworkID,
 			tname: tEvent.TableName,
 			key:   tEvent.Key,
-			node:  nDB.config.NodeName,
+			node:  tEvent.NodeName,
```
Comment (mavenugo) [inline networkdb/delegate.go]: Being a rebroadcast event and based on the current logic of depending on the residual reaptimer, what is the purpose of this change ?

File: networkdb/networkdb.go [hunk 5]
```diff
@@ -558,14 +555,23 @@ func (nDB *NetworkDB) JoinNetwork(nid string) error {
...
+	nodeNetworks[nid] = &network{
+		id:    nid,
+		ltime: ltime,
+		tableBroadcasts: &memberlist.TransmitLimitedQueue{
+			NumNodes: func() int {
+				nDB.RLock()
+				defer nDB.RUnlock()
+				return len(nDB.networkNodes[nid])
+			},
+			RetransmitMult: 4,
```
Comment (mavenugo) [inline networkdb/networkdb.go]: Nit: Pls reorganize this code so that the changes doesnt popup. The only real change in this blob is the addition of `entriesNumber` and the rest of if is purely code reorganization. It is very distracting to see such changes in a PR with other sensitive changes.

File: networkdb/delegate.go (hunk 4 above)
Comment (vieux) [reply networkdb/delegate.go]: if I understood correctly flavio's explanation, it's not strictly related to this PR. it's an issue he saw while he was looking at the code.

Without this change, when the event was rebroadcasted, the node original node emitting the event was replaced with the one rebroadcasting it. So you could receive event saying it originated from one node, but actually came from another one.

File: networkdb/cluster.go (hunk 1 above)
Comment (mavenugo) [reply networkdb/cluster.go]: Yes. this is the current expectation in the code and as @fcrisciani mentioned it is all organized now under the same functions and this assumption is valid.

File: networkdb/networkdb.go [hunk 6]
```diff
@@ -141,6 +141,9 @@ type network struct {
 
 	// Number of gossip messages sent related to this network during the last stats collection period
 	qMessagesSent int
+
+	// Number of entries on the network
+	entriesNumber int
```
Comment (mavenugo) [inline networkdb/networkdb.go]: I guess this is used only for debugging purposes... It would be good to write a comment reflecting the purpose and we should not be tempted to use this as a reference counter for any logic in the future.

**Comment by mavenugo** [comment]
@fcrisciani overall the changes LGTM. Only a few minor questions/comments.

File: networkdb/delegate.go (hunk 3 above)
Comment (fcrisciani) [reply networkdb/delegate.go]: @mavenugo your point is correct only for the case of the truncation of the number. I will add an extra check that the number truncated is still strictly greater than 0.
The improvement here is that the lock is taken only 1 time and not following the same pattern of saving the keys to delete and then do another semaphore race to delete them. So right now the only case in this patch that turns out to be 0 is when the residual time in seconds is 0.xxxx that truncated will be 0.

File: networkdb/delegate.go (hunk 4 above)
Comment (fcrisciani) [reply networkdb/delegate.go]: The node is simply rebroadcasting the message of another node, at the top of the function there is the check that skips messages where the owner matches with the current node.
Without this change the ownership of the entry is messed up and can cause issues on deletion

File: networkdb/networkdb.go (hunk 6 above)
Comment (fcrisciani) [reply networkdb/networkdb.go]: will add extra comments

**Comment by fcrisciani** [comment]
@mavenugo @vieux addressed all the comments, can you guys have a new pass on it?
I will start in the mean time some testing on the patch

File: networkdb/delegate.go (hunk 4 above)
Comment (mavenugo) [reply networkdb/delegate.go]: ok. this is a bit subtle and I think a good testing coverage will help mitigate the risk.

**Comment by vieux** [comment]
New commit LGTM
